prog/mutation.go

```diff
@@ -419,49 +466,128 @@ type mutationArgs struct {
 	target        *Target
 	args          []Arg
 	ctxes         []ArgCtx
+	priorities    []float64
 	ignoreSpecial bool
 }
 
+const (
+	maxPriority = float64(10)
+	minPriority = float64(1)
+)
+
 func (ma *mutationArgs) collectArg(arg Arg, ctx *ArgCtx) {
 	ignoreSpecial := ma.ignoreSpecial
 	ma.ignoreSpecial = false
+
 	switch typ := arg.Type().(type) {
-	case *StructType:
+	case *StructType, *UnionType:
 		if ma.target.SpecialTypes[typ.Name()] == nil || ignoreSpecial {
-			return // For structs only individual fields are updated.
-		}
-		// These special structs are mutated as a whole.
-		ctx.Stop = true
-	case *UnionType:
-		if ma.target.SpecialTypes[typ.Name()] == nil && len(typ.Fields) == 1 || ignoreSpecial {
 			return
 		}
-		ctx.Stop = true
-	case *ArrayType:
-		// Don't mutate fixed-size arrays.
-		if typ.Kind == ArrayRangeLen && typ.RangeBegin == typ.RangeEnd {
-			return
-		}
-	case *CsumType:
-		return // Checksum is updated when the checksummed data changes.
-	case *ConstType:
-		return // Well, this is const.
-	case *BufferType:
-		if typ.Kind == BufferString && len(typ.Values) == 1 {
-			return // string const
-		}
 	case *PtrType:
 		if arg.(*PointerArg).IsSpecial() {
 			// TODO: we ought to mutate this, but we don't have code for this yet.
 			return
 		}
 	}
+
+	prio, flagStop, canMutate := arg.Type().getMutationPrio()
+	ctx.Stop = flagStop
+
+	if !canMutate {
+		return
+	}
+
 	typ := arg.Type()
 	if typ == nil || typ.Dir() == DirOut || !typ.Varlen() && typ.Size() == 0 {
 		return
 	}
+
 	ma.args = append(ma.args, arg)
 	ma.ctxes = append(ma.ctxes, *ctx)
+	ma.priorities = append(ma.priorities, prio)
+}
+
+// TODO: find a way to estimate optimal priority values.
+// Assign a priority for each type. The boolean is the reference type and it has
+// the minimum priority, since it has only two possible values.
+func (t *IntType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	if (t.RangeEnd - t.RangeBegin) == 1 {
+		return minPriority, true, true
+	}
+
+	bitsCount := math.Log2((float64(t.TypeSize * 8)))
+	if t.RangeBegin == t.RangeEnd {
+		return bitsCount + 2, true, true
+	}
+
+	// an integer with a range could have a higher impact
+	return bitsCount + 4, true, true
+}
+
+func (t *StructType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return maxPriority, true, true
+}
+
+func (t *UnionType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	if len(t.Fields) == 1 {
```

Comment:
The table is correct. I didn't notice that in the current code is an 'AND' in that if statement (` ma.target.SpecialTypes[typ.Name()] == nil && len(typ.Fields) == 1 `).  However, I am not sure why the current code has  that check for the number of fields. We only want to mutate structs and unions only when they are special types and otherwise we recursively mutate the fields. Is that correct?